### PR TITLE
 hallucination classifier and helper fixes

### PR DESCRIPTION
**Pull request by Darkxzie**
## Description
- merged the classifier dataset prediction path into `predict_probability(dataset, threshold=None)` so it now returns `(labels, probabilities)`
- added optional threshold override support for single-reaction prediction paths
- normalized ML hallucination helper output so single-string pathways are returned as one-element lists
Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepretro/models/hallucination_classifier.py [hunk 1]
```diff
@@ -271,49 +284,45 @@ def evaluate(self, test_dataset: Dataset, metrics=None) -> dict[str, float]:
...
     def predict_single(
-        self, product_smiles: str, reactants_smiles: str
+        self,
+        product_smiles: str,
+        reactants_smiles: str,
+        threshold: float | None = None,
     ) -> dict[str, Any]:
         """Thin wrapper around :func:`predict_single_reaction`."""
-        return predict_single_reaction(self, product_smiles, reactants_smiles)
+        return predict_single_reaction(
```
Comment (shreyasvinaya) [inline deepretro/models/hallucination_classifier.py]: you might want to remove this function in its entirety

Comment (Darkxzie) [reply deepretro/models/hallucination_classifier.py]: had to add this cuz hallucination_helpers.py calls a predict_single and i have a few doubts on how to change the logic in that file so had to add this.

File: deepretro/models/hallucination_helpers.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    Examples
+    --------
+    >>> from deepretro.models import HallucinationClassifier  # doctest: +SKIP
+    >>> clf = HallucinationClassifier()                       # doctest: +SKIP
+    >>> clf.load("model_out/")                                # doctest: +SKIP
+    >>> checker = build_ml_checker(clf)                       # doctest: +SKIP
+    >>> status, kept = checker("CCO", [["CC", "O"]])          # doctest: +SKIP
+    """
+    def _checker(product: str, pathways: list) -> tuple[int, list]:
```
Comment (riya-singh28) [inline deepretro/models/hallucination_helpers.py]: My suggestion is to change this to a simple MLChecker class for now. It resolves issues with the nested functions and provides a clean pathway for future sub-classes, in case multiple types of MLClassifiers are added. 

```
Class MLChecker:
    
    Init(clf):
        store classifier
        load is_valid_smiles utility

    Call(product, pathways):
        .
        .
        If valid_pathways not empty:
            return (200, valid_pathways)
        Else:
            return (400, [])
```

File: deepretro/models/hallucination_helpers.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            clf = HallucinationClassifier()
+            clf.load(str(classifier))
+        elif hasattr(classifier, "predict_single"):
+            clf = classifier
+        else:
+            raise ValueError(
+                f"hallucination_mode='ml' requires a HallucinationClassifier "
+                f"or path to saved model — got {type(classifier)}"
+            )
+        return build_ml_checker(clf)
```
Comment (riya-singh28) [inline deepretro/models/hallucination_helpers.py]: Please update this according to the suggested new class.

File: deepretro/models/hallucination_helpers.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        *classifier* is not a valid path or model instance.
+
+    Examples
+    --------
+    >>> resolve_hallucination("none", None) is None
+    True
+    >>> checker = resolve_hallucination("heuristic", None)  # doctest: +SKIP
+    >>> callable(checker)                                    # doctest: +SKIP
+    True
+    >>> checker = resolve_hallucination("ml", "model_out/") # doctest: +SKIP
```
Comment (riya-singh28) [inline deepretro/models/hallucination_helpers.py]: Fix doctest, remove skip, add import statements

**Review by riya-singh28 — CHANGES_REQUESTED** [review]
@Darkxzie I have left a few comments after my first review.

File: deepretro/models/hallucination_helpers.py (hunk 2 above)
Comment (Darkxzie) [reply deepretro/models/hallucination_helpers.py]: done.

File: deepretro/models/hallucination_helpers.py (hunk 3 above)
Comment (Darkxzie) [reply deepretro/models/hallucination_helpers.py]: done.

File: deepretro/models/hallucination_helpers.py (hunk 4 above)
Comment (Darkxzie) [reply deepretro/models/hallucination_helpers.py]: done.

File: deepretro/models/hallucination_classifier.py (hunk 1 above)
Comment (riya-singh28) [reply deepretro/models/hallucination_classifier.py]: Please resolve this.

File: deepretro/tests/test_ml_classifier.py [hunk 5]
```diff
@@ -92,9 +93,42 @@ def test_evaluate_keys_and_scores(trained_clf, toy_dataset):
...
-    probability = trained_clf.predict_probability(toy_dataset)
-    assert probability.shape == (len(toy_dataset),)
-    assert np.all((probability >= 0.0) & (probability <= 1.0))
+    labels, probabilities = trained_clf.predict_probability(toy_dataset)
+    assert labels.shape == (len(toy_dataset),)
+    assert probabilities.shape == (len(toy_dataset),)
+    assert np.all((probabilities >= 0.0) & (probabilities <= 1.0))
+
+
+def test_predict_probability_override_uses_explicit_threshold():
```
Comment (riya-singh28) [inline deepretro/tests/test_ml_classifier.py:107]: Add docstrings for all the unit tests

**Review by riya-singh28 — CHANGES_REQUESTED** [review]
I have requested a few changes.

File: deepretro/models/hallucination_classifier.py (hunk 1 above)
Comment (Darkxzie) [reply deepretro/models/hallucination_classifier.py]: resolved.

File: deepretro/models/hallucination_helpers.py
Comment (riya-singh28) [inline deepretro/models/hallucination_helpers.py:1]: @Darkxzie Please add a test file for hallucination_helpers.py and include tests for the resolve hallucination function with model types = heuristic, ml, and none, among other required tests.

**Comment by riya-singh28** [comment]
@Darkxzie Can you also check why these tests are failing? I have not seen it failing for other PRs yet.
```
FAILED test_adv_prompt.py::test_claude_adv_success - assert 400 == 200
FAILED test_llm.py::test_call_llm_success - assert 400 == 200
```

**Comment by Darkxzie** [comment]
> @Darkxzie Can you also check why these tests are failing? I have not seen it failing for other PRs yet.
> 
> ```
> FAILED test_adv_prompt.py::test_claude_adv_success - assert 400 == 200
> FAILED test_llm.py::test_call_llm_success - assert 400 == 200
> ```



> @Darkxzie Can you also check why these tests are failing? I have not seen it failing for other PRs yet.
> 
> ```
> FAILED test_adv_prompt.py::test_claude_adv_success - assert 400 == 200
> FAILED test_llm.py::test_call_llm_success - assert 400 == 200
> ```

Hey riya, I checked the failing run. This pr is from a fork branch, so github actions is running without the anthropic api key which is why both test_claud_adv_success and test_call_llm_success are failing. I did this because i currently do not have permission to create branches in this repo.

**Review by riya-singh28 — APPROVED** [review]
LGTM

**Comment by shreyasvinaya** [comment]
@Darkxzie can you rebase this with main once, the CI should turn green that time, else you can put up a small patch to pin litellm/langfuse (check which is the root cause of the issue)
